lib/charms/ceph_nfs_client/v0/ceph_nfs_client.py

```diff
@@ -0,0 +1,211 @@
+"""CephNfsProvides and CephNfsRequires module.
+
+This library contains the Provides and Requires classes for handling the
+ceph-nfs-client interface.
+
+Import `CephNfsRequires` in your charm, with the charm object and the relation
+name:
+    - self
+    - "ceph-nfs"
+
+Two events are also available to respond to:
+    - connected
+    - departed
+
+A basic example showing the usage of this relation follows:
+
+```
+import charms.ceph_nfs.v0.ceph_nfs as ceph_nfs
+
+
+class CephNfsClientCharm(CharmBase):
+    def __init__(self, *args):
+        super().__init__(*args)
+        # CephNfsRequires
+        self._ceph_nfs = ceph_nfs.CephNfsRequires(
+            self, "ceph-nfs",
+        )
+        self.framework.observe(
+            self._ceph_nfs.on.connected,
+            self._on_ceph_nfs_connected,
+        )
+        self.framework.observe(
+            self._ceph_nfs.on.departed,
+            self._on_ceph_nfs_departed,
+        )
+
+    def _on_ceph_nfs_connected(self, event):
+        '''React to the CephNfsConnectedEvent event.
+
+        This event happens when the ceph-nfs relation is added to the
+        model before information has been provided.
+        '''
+        # Do something before the relation is complete.
+        pass
+
+    def _on_ceph_nfs_departed(self, event):
+        '''React to the CephNfsDepartedEvent event.
+
+        This event happens when ceph-nfs relation is removed.
+        '''
+        # ceph-nfs relation has departed. Shutdown services if needed.
+        pass
+```
+"""
+
+import logging
+
+from ops.charm import (
+    CharmBase,
+    RelationBrokenEvent,
+    RelationChangedEvent,
+    RelationDepartedEvent,
+    RelationEvent,
+)
+from ops.framework import (
+    EventSource,
+    Object,
+    ObjectEvents,
+)
+
+# The unique Charmhub library identifier, never change it
+LIBID = "0fb2d24550d94d24868d3cefa784d5be"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 1
+
+logger = logging.getLogger(__name__)
+
+
+class CephNfsConnectedEvent(RelationEvent):
+    """ceph-nfs connected event."""
+
+    pass
+
+
+class CephNfsDepartedEvent(RelationEvent):
+    """ceph-nfs relation departed event."""
+
+    pass
+
+
+class CephNfsReconcileEvent(RelationEvent):
+    """ceph-nfs relation reconciliation event."""
+
+    pass
+
+
+class CephNfsProvidesEvents(ObjectEvents):
+    """Events class for `on`."""
+
+    ceph_nfs_connected = EventSource(CephNfsConnectedEvent)
+    ceph_nfs_departed = EventSource(CephNfsDepartedEvent)
+    ceph_nfs_reconcile = EventSource(CephNfsReconcileEvent)
+
+
+class CephNfsRequiresEvents(ObjectEvents):
+    """Events class for `on`."""
+
+    ceph_nfs_connected = EventSource(CephNfsConnectedEvent)
+    ceph_nfs_departed = EventSource(CephNfsDepartedEvent)
+
+
+class CephNfsProvides(Object):
```

Comment:
This class defines microceph specific implementation for provider, I am not sure if exposing this in the charm lib has any advantages. Moreover, we end up moving core logic out into a charm lib. Can we please keep the charm lib to only what an external charm (say manila) needs ?

Reply:
Done